enhancements/olm-admission-webhooks.md

```diff
@@ -114,6 +114,27 @@ The certificate creation and lifecycle management is planned to utilize the same
 
 There is a great deal of community support for integration with certificate management projects such as the [service-ca operator](https://github.com/openshift/service-ca-operator) or [cert-manager](https://github.com/jetstack/cert-manager). Cluster Admins should be given the opportunity to integrate with either of these tools be disabling the OLM management of certs and adding appropriate annotations to the deployment and webhook resources via the [SubscriptionConfig resource](https://github.com/operator-framework/operator-lifecycle-manager/blob/8985872bfd5888c8d7d49e4dfc9be162a87691e7/pkg/api/apis/operators/v1alpha1/subscription_types.go#L41-L87).
 
+### Deployment Installation Order
+
+There are a number of usecases in which the order that deployments are rollout matters. Consider the following usecases:
+
+- An operator may be written to assume that any CR it sees has passed through a webhook defined in its CSV. In this case, OLM should provide the Operator Author with the ability to install the admission webhook before the operator to prevent the operator from reconciling CRs that haven't gone through the webhook.
+- An operator may be written such that it expects to run some custom logic prior to starting the webhook. This logic may involve creating a CR that has special values that would be changed by the webhook defined in the CSV.
+
+Both of the usecases described above can be addressed by updating OLM to:
+
+- Create Deployments defined in the CSV in order.
+- Waiting for a Deployment to reach the `READY` state prior to rolling out the next deployment defined in the CSV.
+- Creating the `validatingwebhookconfiguration` or `mutatingwebhookconfiguration` after the deployment has been created.
```

Comment:
This order does not address the first point.

Consider two deployments:

1 - the operator
2 - the admission webhook

If we roll out (1), wait for it to become ready, and then roll out (2) and the validating admission webhooks, (1) may start processing CRs with the assumption that they have passed through a webhook that has not yet been installed.

There may also be inter-dependent operator deployments (that may not need webhooks at all) that can't come into the READY state serially (they may have some coordination to do between themselves before reporting ready). This is not a common case by any means, but it's certainly possible to write a set of deployments with that behavior.

Reply:
Also just to highlight - CRs can exist for the operator before installation at all. This is less common on a new install (though still possible - consider a gitops flow that preloads CRDs/CRs for the operator), but is very common in the upgrade case.

Reply:
> This order does not address the first point.
Consider two deployments:
1 - the operator
2 - the admission webhook
If we roll out (1), wait for it to become ready, and then roll out (2) and the validating admission webhooks, (1) may start processing CRs with the assumption that they have passed through a webhook that has not yet been installed.

@ecordell Is there a reason the operator author shouldn't be responsible for defining the webhook in the deployment list first?

> There may also be inter-dependent operator deployments (that may not need webhooks at all) that can't come into the READY state serially (they may have some coordination to do between themselves before reporting ready). This is not a common case by any means, but it's certainly possible to write a set of deployments with that behavior.

This is a valid point. We can chose to support that usecase if you believe it is important, likely by adding a `priority` field to the Deployment. However, there is an argument that any number of resources should be allowed to declare the order in which they are installed for OLM. I'm slowly becoming convinced that the requirements for operator authors to define the order in which their resources are installed is in fact its own enhancement and I'm hesitant to implement the ability for operator authors to define their deployment installation order without considering how we might order the installation of other resources defined within the CSV.

Reply:
> Is there a reason the operator author shouldn't be responsible for defining the webhook in the deployment list first?

The webhook doesn't become active until the webhook config object is created, so the order of the deployments doesn't matter here. What matters is the ordering between the deployments and the webhook config.

Reply:
There might have been an error in communication. I am trying to propose that we update OLM to support the following scenario. Consider the following deployment set defined in a CSV:

1. A Webhook
2. An Operator

I propose on [L128](https://github.com/operator-framework/enhancements/pull/13/files#diff-d2c1c20116cc404dbf6249792a50f573R128) that after 1 is created, the webhook is created prior to creating 2.

Reply:
That will work - thanks for clarifying. Can we update the language in the doc to make that clear as well?

Reply:
Updated!